### DISCUSSION TITLE
Channel 2 not full 0.5 km resolution - problem with AWS?

### DISCUSSION THREAD

**Discussion by aschueth** [blaylockbk/goes2go#60] [Q&A] [answered]
Hi,

I'm new to this package and trying to understand all the components, but from what I understand using the examples provided, it seems like the resolution, specifically in the red channel, is half of what it should be. For instance, in the CONUS dataset, the size of the xarray dataset is 1500 x 2500 for all of the bands, including channel 2, the 'visible red'. According to the beginners guide (https://noaa-goes16.s3.amazonaws.com/Beginners_Guide_to_GOES-R_Series_Data.pdf), channel 2 should have 0.5 km resolution, while channels 1 and 3 have 1 km resolution. I'm wondering, is that how the level 2 is saved on the AWS bucket? Does level 1b have the full resolution?

**Comment by blaylockbk** [comment, marked as answer]
Hi @aschueth, That's a good question.

You're right, the "Red" channel (channel 2) has twice the resolution as the other channels (0.5 km). Where you get that info depends on the product you open...

- The `ABI-L2-MCMIPC` product is the _Multichannel_ Cloud and Moisture Imagery products; all channels are provided in the same file and on the same 1 km grid (even the "red" channel). This is a convientent product to use if you are making RGB images and need all the data on the same grid.
- The `ABI-L2-CMIPC` is the regular Cloud and Moisture Imagery products, each channel is in its own file. If you look at "Red", the data will be on the 0.5 km grid. If you look at other channels, they will be on the 1 km grid.

**Reply by blaylockbk** [reply]
In GOES-2-go...

This shows the multi-channel level 2 product `ABI-L2-MCMIPC`, with all the channels in the same file on the same grid:

```python
from goes2go import GOES
ds = GOES(satellite=18, product="ABI-L2-MCMIPC").nearesttime('2023-01-01 06:00')
```

![image](https://user-images.githubusercontent.com/6249613/231206787-369a9547-412c-4ca3-96b0-ef8835dc19f1.png)

<br><br>

While this shows the regular level 2 cloud and moisture product `ABI-L2-CMIPC`; since each channel is in its own file you have to specify the channel or "band" you want:

> Note: I never liked the name of the "bands" argument...I need to rename that to "channel"

```python
ds2 = GOES(satellite=18, product="ABI-L2-CMIPC", bands=2).nearesttime('2023-01-01 06:00')
```
![image](https://user-images.githubusercontent.com/6249613/231207076-0bc7ab28-c29d-478a-8636-3027d4509d45.png)

**Comment by aschueth** [comment]
Ahh that did it, thank you very much for clarifying the different products available! And if I want full resolution true color RGB, I can oversample bands/channels 1 and 3 and combine it with 2 like Danielle did here?
https://github.com/daniellelosos/True-Color-Image_GOES-R-ABI-L1b-Radiances/blob/master/RGB%20True-Color%20Image%20from%20GOES-R%20ABI%20L1b%20Radiances.ipynb

**Reply by blaylockbk** [reply]
Yeah, you could oversample those bands like that. To me it seems a little funny to oversample 1 and 3 (creates fake data). Reducing the size of channel 2 makes more sense to me.

```python
regrid_red = ds2.coarsen(dict(y=2,x=2)).mean()

# or

regrid_red = ds2.thin(2)
```

But that depends on your application. Whatever way that makes sense to you.

**Reply by aschueth** [reply]
That makes sense, it is a bit awkward that upsampling creates fake data, but the signal is (mostly?) overwhelmed by the red channel, so the fake data doesn't stand out too much. For my case, the detail in the cloud rolls and towering cumulus are relevant, but I suppose I could always plot channel 2 in isolation instead. Anyway, I thought I would share the comparisons for posterity, it seems like the most notable difference is the blue green hue in the cloud shadows primarily due to the blurring of the blue and green channels?
 
Thanks again so much for your help and assistance, I really appreciate it!

![image](https://user-images.githubusercontent.com/21263541/231252273-29b1bdf6-8f5e-4ae6-9efd-0450ad0f48eb.png)
![image](https://user-images.githubusercontent.com/21263541/231252299-c9200a03-1307-4ec9-aaa6-5b03b1348da1.png)

**Reply by blaylockbk** [reply]
That's a really nice comparison. up vs down/sampling doesn't seem to be much of an issue. Thanks for sharing that.

**Reply by millercommamatt** [reply]
> That makes sense, it is a bit awkward that upsampling creates fake data, but the signal is (mostly?) overwhelmed by the red channel, so the fake data doesn't stand out too much. For my case, the detail in the cloud rolls and towering cumulus are relevant, but I suppose I could always plot channel 2 in isolation instead. Anyway, I thought I would share the comparisons for posterity, it seems like the most notable difference is the blue green hue in the cloud shadows primarily due to the blurring of the blue and green channels?
> 
> Thanks again so much for your help and assistance, I really appreciate it!
> 
> ![image](https://user-images.githubusercontent.com/21263541/231252273-29b1bdf6-8f5e-4ae6-9efd-0450ad0f48eb.png) ![image](https://user-images.githubusercontent.com/21263541/231252299-c9200a03-1307-4ec9-aaa6-5b03b1348da1.png)

@aschueth I know this a 2.5 year old post, but do you happen to have a code example of how you put this together? I too have a need to squeeze a bit more resolution out of the RGB composites. Thanks.

**Reply by aschueth** [reply]
Had to dust off the script, but here is the exact code I used for the upsampled plot above:

```
from goes2go.data import goes_latest, goes_nearesttime
from goes2go import GOES
import matplotlib.pyplot as plt
import cartopy.io.shapereader as shpreader
from cartopy.feature import ShapelyFeature,NaturalEarthFeature
from shapely.geometry import Polygon
from shapely.ops import cascaded_union
from cartopy.io.shapereader import Reader
import datetime
import numpy as np
import gc
import cartopy.crs as ccrs
import cartopy

start_time = datetime.datetime(year=2022,month=5,day=23,hour=19,minute=0)
mins=0
time = start_time+datetime.timedelta(minutes=int(mins))
print(time)
lc = ccrs.PlateCarree()#ccrs.LambertConformal(central_longitude=-97.5, standard_parallels=(38.5,38.5))



g16 = GOES(satellite=16, product="ABI-L2-CMIPC", bands=3).nearesttime('2022-05-23 19:00')
g16.rgb.imshow_kwargs['transform'] = goes_latest(satellite="G16", product="ABI").FOV.crs

R = GOES(satellite=16, product="ABI-L2-CMIPC", bands=2).nearesttime('2022-05-23 19:00')
R = np.power(np.clip(R['CMI'].values, 0, 1), 1/2.2)

B = GOES(satellite=16, product="ABI-L2-CMIPC", bands=1).nearesttime('2022-05-23 19:00')
gradx = np.gradient(B.x.values)[0]
newx = np.arange(B.x.values[0]-gradx/2,B.x.values[-1]+gradx/2,gradx/2)
grady = np.gradient(B.y.values)[0]
newy = np.arange(B.y.values[0]-grady/2,B.y.values[-1]+grady/2,grady/2)
B = np.power(np.clip(B['CMI'].interp(y=newy,x=newx, kwargs={"fill_value": "extrapolate"}).values, 0, 1), 1/2.2)

G = GOES(satellite=16, product="ABI-L2-CMIPC", bands=3).nearesttime('2022-05-23 19:00')
kappa0 = G['kappa0'].values
gradx = np.gradient(G.x.values)[0]
newx = np.arange(G.x.values[0]-gradx/2,G.x.values[-1]+gradx/2,gradx/2)
grady = np.gradient(G.y.values)[0]
newy = np.arange(G.y.values[0]-grady/2,G.y.values[-1]+grady/2,grady/2)
G = np.power(np.clip(G['CMI'].interp(y=newy,x=newx, kwargs={"fill_value": "extrapolate"}).values, 0, 1), 1/2.2)
G_true = 0.45 * R + 0.1 * G + 0.45 * B
G_true = np.clip(G_true, 0, 1)

RGB = np.dstack([R, G_true, B])
RGB = np.nan_to_num(RGB)

fig = plt.figure(figsize=(12, 12))

ax1 = plt.subplot(1,1,1,projection=lc)

ax1.set_extent([-104.5, -99, 31, 36], crs=ccrs.PlateCarree())

ax1.imshow(RGB,**g16.rgb.imshow_kwargs)

fname = '/foo/bar/cb_2017_us_county_5m.shp'
shape_feature = ShapelyFeature(Reader(fname).geometries(),ccrs.PlateCarree(), edgecolor='gray')
ax1.add_feature(shape_feature, facecolor='none', linewidth=0.25, linestyle="-")

states_provinces = cartopy.feature.NaturalEarthFeature(category='cultural',name='admin_1_states_provinces_lines',scale='10m',facecolor='none')
ax1.add_feature(states_provinces, edgecolor='black', linewidth=1)

countries = cartopy.feature.NaturalEarthFeature(category='cultural',name='admin_0_boundary_lines_land',scale='10m',facecolor='none')
ax1.add_feature(countries, edgecolor='black', linewidth=1.5)

ax1.set_title(f"{g16.orbital_slot} True Color", loc="left", fontweight="bold",y=0.96,c='white',fontsize=20)
ax1.text(x=0,y=0.94,s=f"Channel 1&3 Upsampled",c='white',fontsize=15,transform=ax1.transAxes)
ax1.set_title(f"{g16.t.dt.strftime('%H:%M UTC %d-%b-%Y').item()}", loc="right",y=0.96,c='white',fontsize=20)

plt.show()
```

**Reply by millercommamatt** [reply]
Fantastic. Thanks a lot.
